File: C++/Heap/1834_Single-Threaded CPU (Google).cpp
class Solution {
public:
    vector<int> getOrder(vector<vector<int>>& tasks) {
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> nxtask;
        
        vector<array<int, 3>> sortedTasks;
        for(int i = 0 ; i< tasks.size(); i++){
            int enqueTime = tasks[i][0];
            int processingtime = tasks[i][1];
            sortedTasks.push_back({enqueTime, processingtime, i});
        }
        
        sort(begin(sortedTasks), end(sortedTasks));
        vector<int> res;
        
        int taskIndex = 0;
        long long currTime= 0;
        
        while(taskIndex < tasks.size() || !nxtask.empty()){
            if (nxtask.empty() && currTime < sortedTasks[taskIndex][0]) {
               
                currTime = sortedTasks[taskIndex][0];
            }
            
            while(taskIndex < tasks.size() && currTime >= sortedTasks[taskIndex][0]){
                nxtask.push({sortedTasks[taskIndex][1], sortedTasks[taskIndex][2]});
                taskIndex++;
            }
            
            auto [processTime, index] = nxtask.top();
            nxtask.pop();

            
            currTime += processTime;
            res.push_back(index);        
            
            }
        
        
        return res;
    }
};

File: Python/Heap/1834_Single-Threaded CPU (Google).py
class Solution:
    def getOrder(self, tasks: List[List[int]]) -> List[int]:
        next_task = []
        res = []
        
        
        sorted_tasks = [(enqueue, process, idx) for idx,(enqueue, process) in enumerate(tasks)]
        
        sorted_tasks.sort()
        
        curr_time = 0
        task_index = 0
        
        while task_index < len(tasks) or next_task:
            if not next_task and curr_time < sorted_tasks[task_index][0]:
                curr_time = sorted_tasks[task_index][0]
            while task_index < len(tasks) and curr_time >= sorted_tasks[task_index][0]:
                _, process_time, index = sorted_tasks[task_index]
                heapq.heappush(next_task, (process_time, index))
                task_index += 1
            process_time, index_res = heapq.heappop(next_task)
            
            curr_time += process_time
            
            res.append(index_res)
            
        return res